crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,394 @@
+                    }
+
+                    if state.end_of_epoch_flag {
+                        // TODOS:
+                        //
+                        // 1. Here we call this once at the end of the epoch. Depending
+                        // on the performance here, we may instead move to running this as
+                        // a separate background task that accumulates on an ongoing basis.
+                        //
+                        // 2. Here was assume for simplicity that the end of epoch checkpoint will always
+                        // be scheduled last. This is true for now, but we should not assume this.
+                        // Can be fixed to spawn a separate process that does a best effort root digest
+                        // accumulation, with retries in case there are missing checkpoint accumulators.
+                        self.accumulate(state)
+                            .await
+                            .expect("Accumulation failed");
+                    }
+                }
+                // schedule new tasks
+                state = self.queue.recv() => match state {
+                    Some(state) => {
+                        debug!(
+                            ?state,
+                            "StateAccumulator: received enqueued state",
+                        );
+
+                        let store = self.store.clone();
+                        let state_clone = state.clone();
+
+                        pending.push_back(spawn_monitored_task!(async move {
+                            let accumulator = hash_checkpoint_effects(state_clone.clone(), store)?;
+                            Ok((accumulator, state_clone))
+                        }));
+                    }
+                    None => panic!("StateAccumulator: all senders have been dropped"),
+                }
+            }
+        }
+    }
+
+    /// Unions all checkpoint accumulators at the end of the epoch to generate the
+    /// root state hash and saves it. This function is guaranteed to be idempotent (despite the
+    /// underlying data structure not being) as long as it is not called in a multi-threaded
+    /// context.
+    async fn accumulate(&self, last_state_of_epoch: State) -> Result<(), TypedStoreError> {
+        let State {
+            epoch,
+            checkpoint_seq_num: last_checkpoint,
+            ..
+        } = last_state_of_epoch;
+
+        if self
+            .store
+            .tables
+            .root_state_hash_by_epoch
+            .contains_key(&epoch)?
+        {
+            return Ok(());
+        }
+
+        let (next_to_accumulate, mut root_state_hash) = self
+            .store
+            .tables
+            .root_state_hash_by_checkpoint
+            .iter()
+            .skip_to_last()
+            .next()
+            .map(|(highest, hash)| (highest.saturating_add(1), hash))
```

Comment:
i think we want a checked_add here - not that we're going to exceed a u64, but if we do, we should panic